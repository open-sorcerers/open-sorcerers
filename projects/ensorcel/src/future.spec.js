import { after, resolve, fork } from "fluture"
import {
  semiauto,
  auto,
  futurizeWithCancel,
  unfuturize,
  tacit,
  FLUTURE_METHOD_ARITIES
} from "./future"

test("futurizeWithCancel", done => {
  const cancel = () => {}
  const futurize = futurizeWithCancel(cancel)
  const easy = () => Promise.resolve(100)
  const future100Maker = futurize(0, easy)
  fork(done)(z => {
    expect(z).toEqual(100)
    done()
  })(future100Maker())
})

test("unfuturize", done => {
  const easy = () => resolve(100)
  const promise100Maker = unfuturize(0, easy)
  promise100Maker()
    .catch(done)
    .then(z => {
      expect(z).toEqual(100)
      done()
    })
})
test("FLUTURE_METHOD_ARITIES", () => {
  expect(FLUTURE_METHOD_ARITIES).toMatchSnapshot()
})
test("mandatory simple currying sucks", done => {
  const shitLater = after(5)("shit")
  const handler = fork(done)(x => {
    expect(x).toEqual("shit")
    done()
  })
  handler(shitLater)
})
test("tacit - manual", done => {
  const shitLater = after(5)("shit")
  const spoon = tacit(3, fork)
  const handler = spoon(done)(x => {
    expect(x).toEqual("shit")
    done()
  })
  handler(shitLater)
})
test("tacit - all", done => {
  const shitLater = after(5)("shit")
  const spoon = tacit(3, fork)
  spoon(
    done,
    x => {
      expect(x).toEqual("shit")
      done()
    },
    shitLater
  )
})

test("tacit - variant", done => {
  const shitLater = after(5)("shit")
  const spoon = tacit(3, fork)
  spoon(done)(x => {
    expect(x).toEqual("shit")
    done()
  }, shitLater)
})

test("tacit - final permutation", done => {
  const shitLater = after(5)("shit")
  const spoon = tacit(3, fork)
  spoon(done, x => {
    expect(x).toEqual("shit")
    done()
  })(shitLater)
})

test("semiauto", done => {
  const shitLater = after(5)("shit")
  const spoon = semiauto("fork", fork)
  spoon(done, x => {
    expect(x).toEqual("shit")
    done()
  })(shitLater)
})

test("auto", done => {
  const shitLater = after(5)("shit")
  const spoon = auto("fork")
  spoon(done, x => {
    expect(x).toEqual("shit")
    done()
  })(shitLater)
})
