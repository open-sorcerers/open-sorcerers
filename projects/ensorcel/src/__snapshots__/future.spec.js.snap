// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FLUTURE_METHOD_ARITIES 1`] = `
Array [
  Array [
    "swap",
    "promise",
  ],
  Array [
    "ap",
    "attempt",
    "attemptP",
    "chain",
    "chainRej",
    "go",
    "isFuture",
    "isNever",
    "map",
    "mapRej",
    "node",
    "pap",
    "reject",
    "resolve",
  ],
  Array [
    "after",
    "alt",
    "and",
    "both",
    "done",
    "encase",
    "encaseP",
    "lastly",
    "parallel",
    "rejectAfter",
    "value",
  ],
  Array [
    "bimap",
    "coalesce",
    "fork",
    "hook",
  ],
  Array [
    "forkCatch",
  ],
]
`;
